### PR TITLE
Add support for runtime fields

### PR DESCRIPTION
**Pull request by javanna**
This PR includes the work that has been done on the runtime fields feature branch until now. The high level tasks are listed in #59332. The tasks that have not yet been completed can be worked on after merging the feature branch to master.

We are adding a new x-pack plugin called `runtime-fields` that plugs in a custom mapper which allows to define runtime fields based on a script.
The changes included in this PR that were made outside of the `x-pack/plugin/runtime-fields` directory are minimal and revolve around 1) making the `ScriptService` available while parsing index mappings so that the scripts associated to runtime fields can be compiled 2) sharing code to manipulate ranges etc. as it can be reused in runtime fields .

### CODE REVIEW

File: x-pack/plugin/runtime-fields/qa/rest/src/yamlRestTest/java/org/elasticsearch/xpack/runtimefields/rest/CoreTestsWithRuntimeFieldsIT.java
Comment (javanna) [inline x-pack/plugin/runtime-fields/qa/rest/src/yamlRestTest/java/org/elasticsearch/xpack/runtimefields/rest/CoreTestsWithRuntimeFieldsIT.java:242]: @nik9000  this does the job but I am not super proud of it. I guess this is caused by our matching pattern `*` but we have no way to say "all indices but the system ones", and I am not even sure that would work either. Thoughts?

Comment (nik9000) [reply x-pack/plugin/runtime-fields/qa/rest/src/yamlRestTest/java/org/elasticsearch/xpack/runtimefields/rest/CoreTestsWithRuntimeFieldsIT.java:242]: What was wrong? The tests were passing for me without it. I don't mind it - I just don't understand it yet.

File: server/src/main/java/org/elasticsearch/index/query/functionscore/DecayFunctionBuilder.java
Comment (nik9000) [inline server/src/main/java/org/elasticsearch/index/query/functionscore/DecayFunctionBuilder.java:213]: I think I forgot to run these down. I'll add it to my list. But I don't think it blocks merging. We may just never need these to be honest.

File: x-pack/plugin/runtime-fields/src/main/plugin-metadata/plugin-security.policy
Comment (nik9000) [inline x-pack/plugin/runtime-fields/src/main/plugin-metadata/plugin-security.policy:8]: I think we can remove this. I'll add it to my list of things today.

Comment (nik9000) [reply x-pack/plugin/runtime-fields/src/main/plugin-metadata/plugin-security.policy:8]: It also shouldn't block merging.

File: x-pack/plugin/runtime-fields/qa/rest/src/yamlRestTest/java/org/elasticsearch/xpack/runtimefields/rest/CoreTestsWithRuntimeFieldsIT.java
Comment (javanna) [reply x-pack/plugin/runtime-fields/qa/rest/src/yamlRestTest/java/org/elasticsearch/xpack/runtimefields/rest/CoreTestsWithRuntimeFieldsIT.java:242]: yea this came from merging master in today.  see https://github.com/elastic/elasticsearch/pull/61776/commits/acfaccb715f714ebe92b478f9dd3cdfbe7b40228

**Review by nik9000 — APPROVED** [review]
I recognize this code!

Like you told me yesterday, most of the stuff outside of the plugin is about plumbing the script service and handling ranges. Thanks to all of your hard work getting the rest of what we needed into master already.

I found a couple of small things to add to my personal TODO list that I'd forgot about but none of those should block merging.

LGTM

File: x-pack/plugin/runtime-fields/qa/rest/src/yamlRestTest/java/org/elasticsearch/xpack/runtimefields/rest/CoreTestsWithRuntimeFieldsIT.java
Comment (nik9000) [reply x-pack/plugin/runtime-fields/qa/rest/src/yamlRestTest/java/org/elasticsearch/xpack/runtimefields/rest/CoreTestsWithRuntimeFieldsIT.java:242]: Gotcha. It makes sense. It's kind of a shame we don't override the internal templates too to prove that all of those also work with runtime fields but its cool.

**Comment by nik9000** [comment]
Looks like there are some core tests failing with runtime fields:
```
08:51:41   2> REPRODUCE WITH: ./gradlew ':x-pack:plugin:runtime-fields:qa:rest:yamlRestTest' --tests "org.elasticsearch.xpack.runtimefields.rest.CoreTestsWithRuntimeFieldsIT.test {yaml=async_search/20-with-poin-in-time/Async search with point in time}" -Dtests.seed=5BF1520839E0DD82 -Dtests.security.manager=true -Dtests.locale=vi -Dtests.timezone=Asia/Anadyr -Druntime.java=11 
```

I'm cool skipping those for now to unblock this.

File: server/src/main/java/org/elasticsearch/index/query/functionscore/DecayFunctionBuilder.java
Comment (javanna) [reply server/src/main/java/org/elasticsearch/index/query/functionscore/DecayFunctionBuilder.java:213]: I don't think it is worth the effort. I reviewed the leftover TODOs and I am fine with this one staying there.

File: x-pack/plugin/runtime-fields/src/main/plugin-metadata/plugin-security.policy
Comment (javanna) [reply x-pack/plugin/runtime-fields/src/main/plugin-metadata/plugin-security.policy:8]: thanks

File: server/src/main/java/org/elasticsearch/index/query/functionscore/DecayFunctionBuilder.java
Comment (nik9000) [reply server/src/main/java/org/elasticsearch/index/query/functionscore/DecayFunctionBuilder.java:213]: ++

**Comment by nik9000** [comment]
> Looks like there are some core tests failing with runtime fields:

I pushed 1ea67d29f66999b0bfde4973bc37dc7d8b026c73 to fix this.

File: x-pack/plugin/runtime-fields/src/main/plugin-metadata/plugin-security.policy
Comment (nik9000) [reply x-pack/plugin/runtime-fields/src/main/plugin-metadata/plugin-security.policy]: I pushed 34e576fd04cdcfa08fea1ddef3a65f34a142917d to drop it.

**Comment by javanna** [comment]
run elasticsearch-ci/docs

**Comment by javanna** [comment]
run elasticsearch-ci/packaging-sample-windows

**Pull request merged**
